### PR TITLE
feat(learn): add fizzbuzz - RosettaCode challenges

### PR DESCRIPTION
**Pull request by anastasiasoenjoto**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #XXXXX

### CODE REVIEW

**Comment by raisedadead** [comment]
Hi can you confirm or link to a discussion about this PR?

**Comment by anastasiasoenjoto** [comment]
@raisedadead Hi this is a new lesson I dont think there has been a discussion

**Comment by raisedadead** [comment]
Thanks for confirming, no worries.

Typically we would want to have a discussion of the challenges being added to the curriculum.

Here is a repo for the discussion: https://github.com/freeCodeCamp/curriculumexpansion

You should create an issue and link this PR there.

**Comment by anastasiasoenjoto** [comment]
@raisedadead Hi so sorry for this. Thanks for letting me know! I just added a new issue.

**Comment by moT01** [comment]
I was under the impression that it was fine to keep adding stuff to the coding interview prep @raisedadead - not so?

**Comment by raisedadead** [comment]
Yes it is fine to add challenges. 

We should still discuss in issue threads on the curriculum expansion repo just to reduce the likelihood of duplication and avoid 'I was working on a similar challenge' type conflicts amongst contributors.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+<section id='tests'>
+
+```yml
+tests:
+  - text: <code>fizzBuzz</code> should be a function.
+    testString: assert(typeof fizzBuzz=='function');
+  - text: <code>fizzBuzz()</code> should return an Array.
+    testString: assert(Array.isArray(fizzBuzz())==true);
+  - text: Number only divisible by 3 should return <code>"Fizz"</code>.
+    testString: assert.equal(fizzBuzz()[50], "Fizz");
```
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:36]: We can check one of the earlier values here to be sure that the values in the array go from 1 to 100:

`testString: assert.equal(fizzBuzz()[2], "Fizz");`

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  - text: Numbers only divisible by 5 should return <code>"Buzz"</code>.
+    testString: assert.equal(fizzBuzz()[49], "Buzz");
```
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:38]: Similarly, we can check the last value with this test to be sure that the values in the array go up to 100:

`testString: assert.equal(fizzBuzz()[99], "Buzz");`

**Review by scissorsneedfoodtoo — COMMENTED** [review]
@anastasiasoenjoto, thanks again for this PR. I left some thoughts here as comments. Just let me know what you think.

**Comment by anastasiasoenjoto** [comment]
@scissorsneedfoodtoo Thank you for reviewing. I updated the testStrings already let me know if there is anything else I should change.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+## Description
+<section id='description'>
+
+Write a program that outputs the integers from 1 to 100 (inclusive). But: <br>
+<ul>
+    <li> for multiples of 3, output <b>Fizz</b> instead of the number </li>
+    <li> for multiples of 5, output <b>Buzz</b> instead of the number </li>
+    <li> For multiples of 3 and 5, output <b>FizzBuzz</b> instead of the number </li>
+</ul>
```
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:15]: Sorry, I'm not sure what happened to my comment here from yesterday.

Let's clean up some of the spacing here, add code blocks and quotation marks around the strings that should be returned, and change the text slightly:

```
Write a program that returns an array of integers from 1 to 100 (inclusive). But:
<ul>
    <li>for multiples of 3, add <code>"Fizz"</code> to the array instead of the number</li>
    <li>for multiples of 5, add <code>"Buzz"</code> to the array instead of the number</li>
    <li>for multiples of 3 and 5, add <code>"FizzBuzz"</code> to the array instead of the number</li>
</ul>
```

**Comment by scissorsneedfoodtoo** [comment]
@thecodingaviator, could you also check out this challenge and leave a review if you notice anything?

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
+---
+title: FizzBuzz
+id: 5e9ddb06ec35240f39657419
+challengeType: 5
+---
+
+## Description
+<section id='description'>
+
+Write a program that outputs the integers from 1 to 100 (inclusive). But: <br>
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:10]:
```suggestion
Write a program that generates the integers from 1 to 100 (inclusive). But: <br>
```
Why this change? We're returning an array. `outputs` just sounded weird to me

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    <li> For multiples of 3 and 5, output <b>FizzBuzz</b> instead of the number </li>
+</ul>
+
+</section>
+
+## Instructions
+<section id='instructions'>
+
+Your program should return an array containing the results based on the rules above.
+
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:23]:
```suggestion
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    <li> for multiples of 3, output <b>Fizz</b> instead of the number </li>
+    <li> for multiples of 5, output <b>Buzz</b> instead of the number </li>
+    <li> For multiples of 3 and 5, output <b>FizzBuzz</b> instead of the number </li>
+</ul>
+
+</section>
+
+## Instructions
+<section id='instructions'>
+
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:21]:
```suggestion
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+## Tests
+<section id='tests'>
+
+```yml
+tests:
+  - text: <code>fizzBuzz</code> should be a function.
+    testString: assert(typeof fizzBuzz=='function');
+  - text: <code>fizzBuzz()</code> should return an Array.
+    testString: assert(Array.isArray(fizzBuzz())==true);
+  - text: Number only divisible by 3 should return <code>"Fizz"</code>.
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:35]:
```suggestion
  - text: Number divisible by only 3 should return <code>"Fizz"</code>.
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    testString: assert.equal(fizzBuzz()[2], "Fizz");
+  - text: Numbers only divisible by 5 should return <code>"Buzz"</code>.
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:37]:
```suggestion
  - text: Numbers divisible by only 5 should return <code>"Buzz"</code>.
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    testString: assert.equal(fizzBuzz()[99], "Buzz");
+  - text: Numbers divisible by 3 and 5 should return <code>"FizzBuzz"</code>.
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:39]:
```suggestion
  - text: Numbers divisible by both 3 and 5 should return <code>"FizzBuzz"</code>.
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    testString: assert.equal(fizzBuzz()[89], "FizzBuzz");
+  - text: Numbers not divisible by 3 or 5 should return the number itself.
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:41]:
```suggestion
  - text: Numbers not divisible by either 3 or 5 should return the number itself.
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    testString: assert.equal(fizzBuzz()[12], 13);
+
```
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:43]: I don't like how we're doing this though. A user can just hard code these values. I have an alternative approach, I'll put it below

**Review by thecodingaviator — CHANGES_REQUESTED** [review]

**Comment by thecodingaviator** [comment]
How about we just make the user write a program that'll return `Fizz` if `n%3===0`, `Buzz` if n%5===0 and both if divisible by both nos.?

Then we can use randomised testing rather than hard-coding test values

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md (hunk 7 above)
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:35]: The text "only divisible by..." should be fine here. Another alternative is "divisible only by..."

Missed this earlier. Should be:

`  - text: Numbers only divisible by 3 should return <code>"Fizz"</code>.`

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md (hunk 8 above)
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:37]: Again, should be "Numbers" rather than "Number".

**Comment by scissorsneedfoodtoo** [comment]
@thecodingaviator, that's a good point. It's definitely possible that a user will just return an array with the correct values in the indices we're testing. I'm in favor of changing the problem and testing to the way you propose. Either that or we provide the limit and check the entire array that's returned with `deepStrictEqual`.

**Comment by thecodingaviator** [comment]
@scissorsneedfoodtoo I think the best course of action right now is that the OP goes ahead with this challenge then I'll create a PR to implement it in the way I'm proposing. How does that sound?

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md (hunk 7 above)
Comment (thecodingaviator) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:35]: Made the change! Good catch!

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        else {
+            res.push(i);
+        }
+    }
+    return res;
+}
+
+
+
+
```
Comment (moT01) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:89]:
```suggestion
```

Not sure why we need all these extra lines. I guess it works with them but might as well condense a little.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+</div>
+
+</section>
+
+## Solution
+<section id='solution'>
+
+
```
Comment (moT01) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/fizzbuzz.english.md:66]:
```suggestion
```

**Review by moT01 — CHANGES_REQUESTED** [review]
I made a guide page for this challenge...
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-fizzbuzz/385370

You should add the `forumTopicId` to the frontmatter at the top of the file.

**Note to self: Add this solution to the guide page when this gets merged.**

**Review by thecodingaviator — APPROVED** [review]

**Comment by thecodingaviator** [comment]
Actually, I think I know why this wasn't added to the curriculum (if it was intentional). [This](https://www.freecodecamp.org/learn/coding-interview-prep/project-euler/problem-1-multiples-of-3-and-5) challenge takes the exact same skill and tests it, just in a different manner. I really don't think this PR is actually even needed.

@scissorsneedfoodtoo @moT01 thoughts?

**Review by thecodingaviator — CHANGES_REQUESTED** [review]
Check above comment

**Comment by moT01** [comment]
I don't mind adding this even though it is quite similar to the one you linked @thecodingaviator. Many of the challenges in the coding interview area use similar/repeated skills to come up with algorithms. It's more content for users and more repetition to harden those skills.

**Review by moT01 — APPROVED** [review]
LGTM @anastasiasoenjoto 🎉

I will not merge this until someone else can provide input on @thecodingaviator's concern. Remind me to add this solution to the hint article if this gets merged.

**Comment by scissorsneedfoodtoo** [comment]
Thanks for updating your PR @anastasiasoenjoto. Everything LGTM :+1: :+1: 

I agree with @moT01 -- while this challenge is similar to the first Project Euler challenge, I think it's different enough to be included. We already have a more complicated version of FizzBuzz in the Rosetta Code section, and I think this will be a good compliment to that.

It's true that someone could hard code an array with all the values we're checking for, but that would take way longer than actually writing the code to pass the challenge. Still, feel free to update the challenge with another PR @thecodingaviator.

**Review by thecodingaviator — APPROVED** [review]

**Comment by moT01** [comment]
🎉 Thanks for contributing to freeCodeCamp @anastasiasoenjoto 🎉
